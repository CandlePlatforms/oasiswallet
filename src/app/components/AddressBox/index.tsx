/**
 *
 * AddressBox
 *
 */
import copy from 'copy-to-clipboard'
import { Box, Button, Text } from 'grommet'
import { Copy } from 'grommet-icons/icons'
import React, { memo } from 'react'

import { PrettyAddress } from '../PrettyAddress'

interface Props {
  address: string
}

export const AddressBox = memo((props: Props) => {
  const address = props.address

  const copyAddress = () => {
    copy(address)
  }

  return (
    <Box
      direction="row"
      align="center"
      round="5px"
      border={{ color: 'brand' }}
      pad={{ right: 'small' }}
      width="fit-content"
    >
      <Button onClick={() => copyAddress()} icon={<Copy size="18px" />} />
      <Text weight="bold" size="medium" wordBreak="break-word">
        <PrettyAddress address={address} />
      </Text>
    </Box>
  )
})
