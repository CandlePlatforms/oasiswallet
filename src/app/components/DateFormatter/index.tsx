/**
 *
 * DateFormatter
 *
 */
import * as React from 'react'

interface Props {
  date: Date | number
}

export const dateFormat = new Intl.DateTimeFormat(undefined, {
  year: 'numeric',
  month: 'numeric',
  day: 'numeric',
  hour: 'numeric',
  minute: 'numeric',
  second: 'numeric',
  timeZone: process?.env?.NODE_ENV === 'test' ? 'UTC' : undefined,
})

export function DateFormatter(props: Props) {
  const date = props.date instanceof Date ? props.date : props.date * 1000
  const formatted = dateFormat.format(date)
  return <>{formatted}</>
}
