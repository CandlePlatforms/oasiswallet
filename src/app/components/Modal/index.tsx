import { createContext, useCallback, useContext, useState } from 'react'
import { Box, Button, Layer, Heading, Paragraph } from 'grommet'
import { useTranslation } from 'react-i18next'
import { Alert, Checkmark, Close } from 'grommet-icons/icons'

interface Modal {
  title: string
  description: string
  handleConfirm: () => void
  isDangerous: boolean
}

interface ModalContainerProps {
  modal: Modal
  closeModal: () => void
}

interface ModalContextProps {
  launchModal: (modal: Modal) => void
  closeModal: () => void
}

interface ModalProviderProps {
  children: React.ReactNode
}

const ModalContext = createContext<ModalContextProps>({} as ModalContextProps)

const ModalContainer = ({ modal, closeModal }: ModalContainerProps) => {
  const { t } = useTranslation()
  const confirm = useCallback(() => {
    modal.handleConfirm()
    closeModal()
  }, [closeModal, modal])

  return (
    <Layer onEsc={closeModal} onClickOutside={closeModal} background="background-front">
      <Box margin="medium">
        <Heading size="small">{modal.title}</Heading>
        <Paragraph fill>{modal.description}</Paragraph>
        <Box direction="row" gap="small" alignSelf="end" pad={{ top: 'large' }}>
          <Button
            label={t('common.cancel', 'Cancel')}
            onClick={closeModal}
            secondary
            style={{ borderRadius: '4px' }}
            icon={<Close size="18px" />}
          />
          <Button
            label={t('common.confirm', 'Confirm')}
            onClick={confirm}
            disabled={modal.isDangerous}
            primary={modal.isDangerous}
            color={modal.isDangerous ? 'status-error' : ''}
            style={{ borderRadius: '4px' }}
            icon={modal.isDangerous ? <Alert size="18px" /> : <Checkmark size="18px" />}
          />
        </Box>
      </Box>
    </Layer>
  )
}

const ModalProvider = (props: ModalProviderProps) => {
  const [modal, setModal] = useState<Modal | null>(null)
  const closeModal = useCallback(() => {
    setModal(null)
  }, [])

  return (
    <ModalContext.Provider value={{ closeModal, launchModal: setModal }}>
      {props.children}
      {modal && <ModalContainer modal={modal} closeModal={closeModal} />}
    </ModalContext.Provider>
  )
}

const useModal = () => {
  const context = useContext(ModalContext)
  if (context === undefined) {
    throw new Error('useModal must be used within a ModalProvider')
  }

  return context
}

export { ModalProvider, useModal }
