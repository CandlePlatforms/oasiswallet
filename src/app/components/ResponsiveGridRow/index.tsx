/**
 *
 * ResponsiveGridRow
 *
 */
import { Box, Text } from 'grommet'
import React, { memo } from 'react'

interface Props {
  label: string | React.ReactNode
  value: string | React.ReactNode
}

export const ResponsiveGridRow = memo((props: Props) => {
  return (
    <>
      <Box>
        <Text weight="bold">{props.label}</Text>
      </Box>
      <Box>{props.value}</Box>
    </>
  )
})
