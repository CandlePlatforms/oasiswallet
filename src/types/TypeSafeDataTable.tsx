import DataTable, { IDataTableColumn, IDataTableProps } from 'react-data-table-component'

export interface ITypeSafeDataTableColumn<T> extends IDataTableColumn<T> {
  id: string
}
export interface ITypeSafeDataTableProps<T> extends Omit<IDataTableProps<T>, 'keyField'> {
  keyField: keyof T
  columns: ITypeSafeDataTableColumn<T>[]
}

/**
 * Overrides DataTable type to ensure:
 * - `keyField` on rows is not missing and is valid
 * - `id` on cols is not missing
 */
export function TypeSafeDataTable<T = any>(props: ITypeSafeDataTableProps<T>): React.ReactElement {
  return <DataTable {...(props as IDataTableProps<T>)} />
}
