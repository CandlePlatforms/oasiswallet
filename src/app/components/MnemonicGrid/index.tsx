import { NoTranslate } from 'app/components/NoTranslate'
import { Box, Grid, ResponsiveContext, Text } from 'grommet'
import * as React from 'react'
import { useContext } from 'react'

/**
 *
 * MnemonicGrid
 *
 */
interface WordProp {
  id: number
  word: string
  hidden?: boolean
  higlighted?: boolean
}

let noSelect: React.CSSProperties = {
  WebkitUserSelect: 'none',
  MozUserSelect: 'none',
  msUserSelect: 'none',
  userSelect: 'none',
}

// Make typos obvious e.g. if user has pasted mnemonic containing a newline
let keepWhitespace: React.CSSProperties = {
  whiteSpace: 'pre',
}

function MnemonicWord(props: WordProp) {
  return (
    <Box
      background={props.higlighted ? 'brand' : 'background-contrast'}
      margin="xsmall"
      direction="row"
      pad="xsmall"
      border={{ side: 'bottom' }}
    >
      <Box pad={{ right: 'small' }}>
        <Text style={noSelect}>{props.id}.</Text>
      </Box>
      <Box>
        <NoTranslate>
          <strong style={keepWhitespace}>{props.hidden ? '' : props.word}</strong>
        </NoTranslate>
      </Box>
    </Box>
  )
}

interface Props {
  // List of words
  mnemonic: string[]

  /** Indexes of hidden words, used for mnemonic validation */
  hiddenWords?: number[]

  /** Highlighted word indexes, used for mnemonic validation */
  highlightedIndex?: number
}

export function MnemonicGrid({ mnemonic, highlightedIndex: hilightedIndex, hiddenWords }: Props) {
  const size = useContext(ResponsiveContext)
  const maxEnglishLength = 8
  const numberDotSpaceLength = 4
  const columnSize =
    size === 'large' ? ['1fr', '1fr', '1fr', '1fr'] : `${maxEnglishLength + numberDotSpaceLength + 2}ch`

  return (
    <NoTranslate>
      <Grid columns={columnSize} data-testid="mnemonic-grid">
        {mnemonic.map((word, index) => (
          <MnemonicWord
            key={index + 1}
            id={index + 1}
            word={word}
            higlighted={index === hilightedIndex}
            hidden={hiddenWords && hiddenWords.indexOf(index) !== -1}
          />
        ))}
      </Grid>
    </NoTranslate>
  )
}
