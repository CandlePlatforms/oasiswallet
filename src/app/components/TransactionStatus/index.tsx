/**
 *
 * TransactionStatus
 *
 */
import { Box, Text } from 'grommet'
import React, { memo } from 'react'
import { useTranslation } from 'react-i18next'
import { ErrorPayload } from 'types/errors'
import { ErrorFormatter } from '../ErrorFormatter'

interface Props {
  error?: ErrorPayload
  success?: boolean
}

export const TransactionStatus = memo((props: Props) => {
  const { error, success } = props
  const { t } = useTranslation()

  return (
    <>
      {error && (
        <Box
          border={{
            color: 'status-error',
            side: 'left',
            size: '3px',
          }}
          background={{
            color: 'status-error',
            opacity: 'weak',
          }}
          pad={{ horizontal: 'small', vertical: 'xsmall' }}
        >
          <Text weight="bold">
            <ErrorFormatter code={error.code} message={error.message} />
          </Text>
        </Box>
      )}
      {success && (
        <Box
          border={{
            color: 'status-ok',
            side: 'left',
            size: '3px',
          }}
          background={{
            color: 'status-ok',
            opacity: 'weak',
          }}
          pad={{ horizontal: 'small', vertical: 'xsmall' }}
        >
          <Text weight="bold">
            {t(
              'account.sendTransaction.success',
              'Transaction successfully sent. The transaction might take up to a minute to appear on your account.',
            )}
          </Text>
        </Box>
      )}
    </>
  )
})
