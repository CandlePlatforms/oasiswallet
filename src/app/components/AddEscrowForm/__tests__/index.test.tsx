import { render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { ModalProvider } from 'app/components/Modal'
import { Validator } from 'app/state/staking/types'
import * as React from 'react'
import { Provider } from 'react-redux'
import { configureAppStore } from 'store/configureStore'

import { AddEscrowForm } from '..'

const renderComponent = (store: any, address: string, validatorStatus: Validator['status']) =>
  render(
    <Provider store={store}>
      <ModalProvider>
        <AddEscrowForm validatorAddress={address} validatorStatus={validatorStatus} />
      </ModalProvider>
    </Provider>,
  )

describe('<AddEscrowForm />', () => {
  let store: ReturnType<typeof configureAppStore>

  beforeEach(() => {
    store = configureAppStore()
  })

  it('should match snapshot', () => {
    const component = renderComponent(store, 'dummy-address', 'active')
    expect(component.container.firstChild).toMatchSnapshot()
  })

  it('should dispatch an addEscrow transaction', () => {
    renderComponent(store, 'dummy-address', 'active')

    userEvent.type(screen.getByTestId('amount'), '1000')
    userEvent.click(screen.getByRole('button'))

    expect(screen.getByText('errors.noOpenWallet')).toBeInTheDocument()
  })

  it('should warn before addEscrow to inactive validator', () => {
    renderComponent(store, 'dummy-address', 'inactive')

    userEvent.type(screen.getByTestId('amount'), '1000')
    userEvent.click(screen.getByRole('button'))

    expect(screen.getByText('account.addEscrow.confirmDelegatingToInactive.title')).toBeInTheDocument()
  })
})
