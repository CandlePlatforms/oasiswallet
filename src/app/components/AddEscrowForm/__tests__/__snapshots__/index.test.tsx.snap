// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddEscrowForm /> should match snapshot 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  min-width: 0;
  min-height: 0;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  padding-top: 12px;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  background-color: #FFFFFF;
  color: #444444;
  min-width: 0;
  min-height: 0;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c4 {
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  -webkit-align-self: stretch;
  -ms-flex-item-align: stretch;
  align-self: stretch;
  width: 12px;
}

.c5 {
  display: inline-block;
  box-sizing: border-box;
  cursor: pointer;
  font: inherit;
  -webkit-text-decoration: none;
  text-decoration: none;
  margin: 0;
  background: transparent;
  overflow: visible;
  text-transform: none;
  border: 2px solid #7D4CDB;
  border-radius: 18px;
  color: #444444;
  padding: 4px 22px;
  font-size: 18px;
  line-height: 24px;
  background-color: #7D4CDB;
  color: #f8f8f8;
  border-radius: 18px;
  -webkit-transition-property: color,background-color,border-color,box-shadow;
  transition-property: color,background-color,border-color,box-shadow;
  -webkit-transition-duration: 0.1s;
  transition-duration: 0.1s;
  -webkit-transition-timing-function: ease-in-out;
  transition-timing-function: ease-in-out;
}

.c5:hover {
  box-shadow: 0px 0px 0px 2px #7D4CDB;
}

.c5:focus {
  outline: none;
  box-shadow: 0 0 2px 2px #6FFFB0;
}

.c5:focus > circle,
.c5:focus > ellipse,
.c5:focus > line,
.c5:focus > path,
.c5:focus > polygon,
.c5:focus > polyline,
.c5:focus > rect {
  outline: none;
  box-shadow: 0 0 2px 2px #6FFFB0;
}

.c5:focus::-moz-focus-inner {
  border: 0;
}

.c5:focus:not(:focus-visible) {
  outline: none;
  box-shadow: none;
}

.c5:focus:not(:focus-visible) > circle,
.c5:focus:not(:focus-visible) > ellipse,
.c5:focus:not(:focus-visible) > line,
.c5:focus:not(:focus-visible) > path,
.c5:focus:not(:focus-visible) > polygon,
.c5:focus:not(:focus-visible) > polyline,
.c5:focus:not(:focus-visible) > rect {
  outline: none;
  box-shadow: none;
}

.c5:focus:not(:focus-visible)::-moz-focus-inner {
  border: 0;
}

.c3 {
  box-sizing: border-box;
  font-size: inherit;
  font-family: inherit;
  border: none;
  -webkit-appearance: none;
  background: transparent;
  color: inherit;
  width: 100%;
  padding: 11px;
  font-weight: 600;
  margin: 0;
  border: 1px solid rgba(0,0,0,0.33);
  border-radius: 4px;
}

.c3:focus {
  outline: none;
  box-shadow: 0 0 2px 2px #6FFFB0;
}

.c3:focus > circle,
.c3:focus > ellipse,
.c3:focus > line,
.c3:focus > path,
.c3:focus > polygon,
.c3:focus > polyline,
.c3:focus > rect {
  outline: none;
  box-shadow: 0 0 2px 2px #6FFFB0;
}

.c3:focus::-moz-focus-inner {
  border: 0;
}

.c3::-webkit-input-placeholder {
  color: #AAAAAA;
}

.c3::-moz-placeholder {
  color: #AAAAAA;
}

.c3:-ms-input-placeholder {
  color: #AAAAAA;
}

.c3::-webkit-search-decoration {
  -webkit-appearance: none;
}

.c3::-moz-focus-inner {
  border: none;
  outline: none;
}

.c3:-moz-placeholder,
.c3::-moz-placeholder {
  opacity: 1;
}

.c2 {
  position: relative;
  width: 100%;
}

@media only screen and (max-width:768px) {
  .c0 {
    padding-top: 6px;
  }
}

@media only screen and (max-width:768px) {
  .c4 {
    width: 6px;
  }
}

<form>
  <div
    class="c0"
  >
    <div
      class="c1"
    >
      <div
        class="c2"
      >
        <input
          autocomplete="off"
          class="c3"
          data-testid="amount"
          id="amount-id"
          min="0"
          name="amount"
          placeholder="common.amount"
          required=""
          step="any"
          type="number"
          value=""
        />
      </div>
    </div>
    <div
      class="c4"
    />
    <button
      class="c5"
      style="border-radius: 4px;"
      type="submit"
    >
      account.addEscrow.delegate
    </button>
  </div>
</form>
`;
